keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,274 @@
+reviewers:
+  - "@saad-ali"
+  - "@alarge"
+  - "@erinboyd"
+  - "@guymguym"
+  - "@travisn"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-25
+last-updated: 2020-02-27
+status: provisional
+---
+
+# Object Bucket Provisioning
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Vocabulary](#vocabulary)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+      - [Admin](#admin)
+      - [User](#user)
+  - [System Configuration](#system-configuration)
+  - [Workflows](#workflows)
+    - [Create Bucket](#create-bucket)
+    - [Delete Bucket](#delete-bucket)
+    - [Grant Bucket Access](#grant-bucket-access)
+    - [Revoke Bucket Access](#revoke-bucket-access)
+  - [Custom Resource Definitions](#custom-resource-definitions)
+      - [ObjectBucketClaim (OBC)](#objectbucketclaim-obc)
+      - [ObjectBucket (OB)](#objectbucket-ob)
+      - [BucketClass](#bucketclass)
+<!-- /toc -->
+
+# Summary
+
+This proposal introduces the Container Object Storage Interface (COSI), whose purpose is to provide a familiar and standardized method of provisioning object storage buckets in an manner agnostic to the storage vendor. The COSI environment is comprised of Kubernetes CRDs, operators to manage these CRDs, and a gRPC interface by which these operators communicate with vendor plugins.  This design is of course heavily inspired by the Kubernetes’ implementation of the Container Storage Interface (CSI).  However, bucket management lacks some of the notable requirements of block and file provisioning and allows for an architecture with lower overall complexity.  This proposal does not include a standardized protocol or abstraction of storage vendor APIs.  
+
+## Motivation 
+
+File and block are first class citizens within the Kubernetes ecosystem.  Object, though different on a fundamental level and lacking an open, committee controlled interface like POSIX, is a popular means of storing data, especially against very large data sources.   As such, we feel it is in the interest of the community to elevate buckets to a community supported feature.  In doing so, we can provide Kubernetes cluster users and administrators a normalized and familiar means of managing object storage.
+
+## Goals
++ Define a control plane API in order to standardize and formalize a Kubernetes bucket provisioning
++ Minimize privileges required to run a storage driver.
++ Minimize technical ramp-up for storage vendors.
++ Be un-opinionated about the underlying object-store.
++ Use Kubernetes Secrets to inject bucket information into application workflows.
++ Present similar workflows for both _greenfield_  and _brownfield_ bucket provisioning.
++ Present a design that is familiar to CSI storage plugin authors and Kubernetes storage admins.
+
+## Non-Goals
++ Define a native _data-plane_ object store API which would greatly improve object store app portability.
```

Comment:
Not sure how to handle this in general. The main difference then between block/filesystem provisioning and this is in block/filesystem the user is abstracted away from knowing what protocol is used to talk to the backing store, how to authenticate with the backing store if required, and even needing any kind of driver at all to talk to it. K8s itself provides the container with just a mount.

If an abstract dataplane is not provided, then all those details must be part of the api itself and left up as an implementation detail to the app, making it potentially not portable.

It still may be worth doing though.

Reply:
This has been debated for a long time, and in fact has a pretty rich startup space, too .  Providing a universal data plane API means either a) choosing a lowest-common API (a subset of S3 seems most widely plausible) or injecting an actor into the datapath (which is scary from an auth and perf and reliability POV).

Alternatively, let this be a client-side thing.  Make client libs that understand backends natively.  Leave this API for "creations, credentials, and coordinates".

Reply:
@kfox1111 +1 As Tim pointed out, the idea has kicked around for a while.  One of the motivations for this KEP is to lay some groundwork for a k8s native object interface.

Reply:
I would like to work on a data plane object KEP, but after this KEP is (hopefully) merged and a code pr is underway.